src/System.Private.CoreLib/shared/System/IO/Stream.cs

```diff
@@ -239,8 +243,12 @@ public override Task WriteAsync(byte[] buffer, int offset, int length, Cancellat
 
             public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
             {
-                if (_func == null) throw new NotSupportedException();
-                return _func(buffer, _state, cancellationToken);
+                if (_func != null)
+                    return _func(buffer, _state, cancellationToken);
+                else if (_action != null)
+                    return new ValueTask(Task.Run(() => _action(buffer.Span, _state), cancellationToken));
```

Comment:
`Task.Run` will cause a closure to be created (even if `_func` is not null). Calling `_action` directly might be enough?

Reply:
Not sure to understand the closure issue… Is it better with the commit below?

Reply:
`return _func` should be left as it was and for `_action` you could use something like this:
```c#
else if (_action != null)
{
    _action(buffer.Span, _state), cancellationToken);
    return default;
}